From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Mon, 26 Apr 2021 11:20:12 -0500
Subject: [PATCH] Reduce frequency of checking for entity despawn

This isn't a great way to deal with this, but at a high number of
entities this starts adding up more than it should, and it's already
fairly optimized. Just starts being slow when you have 30k entities.

diff --git a/src/main/java/gg/airplane/AirplaneConfig.java b/src/main/java/gg/airplane/AirplaneConfig.java
index e22ccbe13a252b9406c02d5c516b245d5ff57a0f..3c13ee6b907add5fbb51f6955020b6bc3ac28c6f 100644
--- a/src/main/java/gg/airplane/AirplaneConfig.java
+++ b/src/main/java/gg/airplane/AirplaneConfig.java
@@ -109,4 +109,13 @@ public class AirplaneConfig {
     }
 
 
+    public static byte entityDespawnCheckFrequency;
+
+    private static void entitySettings() {
+        config.setComment("entities", "Configures settings for generic entities");
+
+        entityDespawnCheckFrequency = (byte) Math.max(config.getInt("entities.despawn-check-freq", 8), Byte.MAX_VALUE);
+    }
+
+
 }
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index b4d23464960fa2a4056d915909c1db0c7650d8f6..9197eb6e3bbb9885a40f56a274eef92dfacd5a20 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -817,8 +817,15 @@ public abstract class Mob extends LivingEntity {
         return false;
     }
 
+    // Airplane start - just reduce frequency of checking for despawn
+    private byte despawnCheck = 0;
     @Override
     public void checkDespawn() {
+        if (++despawnCheck < gg.airplane.AirplaneConfig.entityDespawnCheckFrequency) {
+            return;
+        }
+        this.despawnCheck = 0;
+        // Airplane end
         if (this.level.getDifficulty() == Difficulty.PEACEFUL && this.shouldDespawnInPeaceful()) {
             this.discard();
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
